Lib/asyncio/tasks.py

```diff
@@ -43,21 +43,8 @@ def all_tasks(loop=None):
     if loop is None:
         loop = events.get_running_loop()
     # Looping over a WeakSet (_all_tasks) isn't safe as it can be updated from another
-    # thread while we do so. Therefore we cast it to list prior to filtering. The list
-    # cast itself requires iteration, so we repeat it several times ignoring
-    # RuntimeErrors (which are not very likely to occur). See issues 34970 and 36607 for
-    # details.
-    i = 0
-    while True:
-        try:
-            tasks = list(_all_tasks)
-        except RuntimeError:
-            i += 1
-            if i >= 1000:
-                raise
-        else:
-            break
-    return {t for t in tasks
+    # thread while we do so. Therefore we copy it prior to filtering.
+    return {t for t in _all_tasks.copy()
             if futures._get_loop(t) is loop and not t.done()}
```

Comment:
I'm happy to see this finally addressed. Note: https://bugs.python.org/issue36607#msg345380

It's a behavioral change though: an unintentional feature of the previous implementation was that, due to restarting on mutation, it returned a set that was "as fresh as possible". Your version might be a tiny bit off. However, as discussion on the issue I linked demonstrates, the previous behavior was only working for a subset of possible cases so in the end I think the atomicity is better.